core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala

```diff
@@ -0,0 +1,210 @@
+    val input = "headerKey:headerValue\tvalue\n"
+
+    val props = defaultTestProps
+    props.put("parse.key", "false")
+
+    runTest(props, input, record(null, "value", List("headerKey" -> "headerValue")))
+  }
+
+  @Test
+  def testLineReaderOnlyValue(): Unit = {
+    val props = defaultTestProps
+    props.put("parse.key", "false")
+    props.put("parse.headers", "false")
+
+    runTest(props, "value\n", record(null, "value"))
+  }
+
+  @Test
+  def testParseHeaderEnabledWithCustomDelimiterAndVaryingNumberOfKeyValueHeaderPairs(): Unit = {
+    val props = defaultTestProps
+    props.put("key.separator", "#")
+    props.put("parse.headers", "true")
+    props.put("headers.delimiter", "!")
+    props.put("headers.separator", "&")
+    props.put("headers.key.separator", ":")
+
+    val input =
+      "headerKey0.0:headerValue0.0&headerKey0.1:headerValue0.1!key0#value0\n" +
+      "headerKey1.0:headerValue1.0!key1#value1"
+
+    val record0 = record("key0", "value0", List("headerKey0.0" -> "headerValue0.0", "headerKey0.1" -> "headerValue0.1"))
+    val record1 = record("key1", "value1", List("headerKey1.0" -> "headerValue1.0"))
+
+    runTest(props, input, record0, record1)
+  }
+
+  @Test
+  def testMissingHeaderSeparator(): Unit = {
+    val lineReader = new LineMessageReader()
+    val input =
+      "headerKey0.0:headerValue0.0,headerKey0.1:headerValue0.1\tkey0\tvalue0\n" +
+      "headerKey1.0:headerValue1.0\tkey1[MISSING-DELIMITER]value1"
+
+    lineReader.init(new ByteArrayInputStream(input.getBytes), defaultTestProps)
+    lineReader.readMessage()
+
+    val expectedException = assertThrows(classOf[KafkaException], () => lineReader.readMessage())
+
+    assertEquals(
+      "No key separator found in 'key1[MISSING-DELIMITER]value1' on line number 2",
+      expectedException.getMessage
+    )
+  }
+
+  @Test
+  def testHeaderDemarcationCollision(): Unit = {
+    val props = defaultTestProps
+    props.put("headers.delimiter", "\t")
+    props.put("headers.separator", "\t")
+    props.put("headers.key.separator", "\t")
+
+    assertThrowsOnInvalidPatternConfig(props, "headers.delimiter and headers.separator may not be equal")
+
+    props.put("headers.separator", ",")
+    assertThrowsOnInvalidPatternConfig(props, "headers.delimiter and headers.key.separator may not be equal")
+
+    props.put("headers.key.separator", ",")
+    assertThrowsOnInvalidPatternConfig(props, "headers.separator and headers.key.separator may not be equal")
+  }
+
+  private def assertThrowsOnInvalidPatternConfig(props: Properties, expectedMessage: String): Unit = {
+    val exception = assertThrows(classOf[KafkaException], () => new LineMessageReader().init(null, props))
+    assertEquals(
+      expectedMessage,
+      exception.getMessage
+    )
+  }
+
+  @Test
+  def testIgnoreErrorInInput(): Unit = {
+    val input =
+      "headerKey0.0:headerValue0.0,headerKey0.1:headerValue0.1[MISSING-DELIMITER]key0\tvalue0\n" +
```

Comment:
Should we also add cases for missing delimiter in the headers part? For instance, a missing `:` or a missing `,` or even both at the same time.